### PR TITLE
Added gRPC for e-mail status events.

### PR DESCRIPTION
**Pull request by robin-wallberg**

### CODE REVIEW

File: tachikoma-frontend-api-proto/src/main/proto/com/sourceforgery/tachikoma/grpc/admin/emailstatusevent/emailstatusevent.proto [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    SOFT_BOUNCED = 2;
+    DELIVERED = 3;
+    OPENED = 4;
+    CLICKED = 5;
+    SPAM = 6;
+    UNSUBSCRIBE = 7;
+}
+
+// An email status event
+message EmailStatusEvent {
```
Comment (robin-wallberg) [inline tachikoma-frontend-api-proto/src/main/proto/com/sourceforgery/tachikoma/grpc/admin/emailstatusevent/emailstatusevent.proto]: Remove, use existing Notification

File: tachikoma-database/src/main/kotlin/com/sourceforgery/tachikoma/database/upgrades/Version3.kt [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+import java.sql.Connection
+
+class Version3 : DatabaseUpgrade {
+    override fun run(connection: Connection): Int {
+        connection
+                .createStatement()
+                .use {
+                    it.executeUpdate("ALTER TABLE e_email_status DROP COLUMN mta_status_code")
+                    it.executeUpdate("ALTER TABLE e_email_status ADD COLUMN metaData JSON")
```
Comment (tobias-) [inline tachikoma-database/src/main/kotlin/com/sourceforgery/tachikoma/database/upgrades/Version3.kt]: Use not null and set default value to '{}'

File: tachikoma-database/src/main/kotlin/com/sourceforgery/tachikoma/database/upgrades/Version3.kt [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+                    it.executeUpdate("CREATE INDEX mail_status_meta_data_mta_status_code ON e_email_status ((metaData->>'mtaStatusCode'))")
+                    it.executeUpdate("CREATE INDEX mail_status_meta_data_ip_address ON e_email_status ((metaData->>'ipAddress'))")
```
Comment (tobias-) [inline tachikoma-database/src/main/kotlin/com/sourceforgery/tachikoma/database/upgrades/Version3.kt]: Remove index until we use them

File: tachikoma-database/src/test/kotlin/com/sourceforgery/tachikoma/database/dao/EmailStatusEventDAOSpec.kt [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    describe("EmailStatusEventDAO") {
+
+        var accountDBO: AccountDBO? = null
+
+        beforeEachTest {
+            accountDBO = AccountDBO(MailDomain("example.com"))
+            accountDAO.save(accountDBO!!)
+        }
+
+        afterEachTest {
```
Comment (tobias-) [inline tachikoma-database/src/test/kotlin/com/sourceforgery/tachikoma/database/dao/EmailStatusEventDAOSpec.kt]: Should not be needed as `serviceLocator.shutdown()` cleans this up

File: tachikoma-frontend-api-proto/src/main/proto/com/sourceforgery/tachikoma/grpc/admin/emailstatusevent/emailstatusevent.proto [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+import "google/protobuf/empty.proto";
+import "google/protobuf/timestamp.proto";
+import "com/sourceforgery/tachikoma/grpc/frontend/common.proto";
+
+package com.sourceforgery.tachikoma.grpc.frontend.emailstatusevent;
+
+service EmailStatusEventService {
+
+    // Returns all email status events the last 7 days
+    rpc getEmailStatusEvents (google.protobuf.Empty) returns (stream EmailNotification);
```
Comment (tobias-) [inline tachikoma-frontend-api-proto/src/main/proto/com/sourceforgery/tachikoma/grpc/admin/emailstatusevent/emailstatusevent.proto]: Add filter object. From the beginning, just filter on days in history

File: tachikoma-grpc/src/main/kotlin/com/sourceforgery/tachikoma/emailstatusevent/EmailStatusEventService.kt [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+private constructor(
+        private val authentication: Authentication,
+        private val emailStatusEventDAO: EmailStatusEventDAO
+) {
+    fun getEmailStatusEvents(responseObserver: StreamObserver<EmailNotification>) {
+
+        authentication.requireFrontend()
+
+        // TODO What is an reasonable amount of time?
+        val timeFromNow = Instant.now().minus(7, ChronoUnit.DAYS)
```
Comment (tobias-) [inline tachikoma-grpc/src/main/kotlin/com/sourceforgery/tachikoma/emailstatusevent/EmailStatusEventService.kt]: Add filter here

File: tachikoma-grpc/src/main/kotlin/com/sourceforgery/tachikoma/emailstatusevent/EmailStatusEventService.kt [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+        emailStatusEventDAO.getEventsAfter(timeFromNow)
+                .forEach {
```
Comment (tobias-) [inline tachikoma-grpc/src/main/kotlin/com/sourceforgery/tachikoma/emailstatusevent/EmailStatusEventService.kt]: Make sure to only return emails sent from this account

File: tachikoma-grpc/src/main/kotlin/com/sourceforgery/tachikoma/emailstatusevent/EmailStatusEventService.kt [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+            EmailStatus.UNSUBSCRIBE -> {
+                builder.setUnsubscribedEvent(UnsubscribedEvent.getDefaultInstance())
+            }
+            EmailStatus.QUEUED -> {
+                builder.setQueuedEvent(QueuedEvent.getDefaultInstance())
+            }
+            EmailStatus.SPAM -> {
+                builder.setSpamEvent(SpamEvent.getDefaultInstance())
+            }
+        }
```
Comment (tobias-) [inline tachikoma-grpc/src/main/kotlin/com/sourceforgery/tachikoma/emailstatusevent/EmailStatusEventService.kt]: add `}.build()`

File: tachikoma-grpc/src/main/kotlin/com/sourceforgery/tachikoma/emailstatusevent/EmailStatusEventService.kt [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+                }
+        responseObserver.onCompleted()
+    }
+
+    private fun getEmailNotification(it: EmailStatusEventDBO): EmailNotification.Builder {
+        val builder = EmailNotification.newBuilder()
+        builder.emailId = it.email.id.toGrpcInternal()
+        builder.recipientEmailAddress = it.email.recipient.toGrpcInternal()
+        builder.emailTransactionId = it.email.transaction.id.toGrpcInternal()
+        builder.timestamp = Timestamp
```
Comment (tobias-) [inline tachikoma-grpc/src/main/kotlin/com/sourceforgery/tachikoma/emailstatusevent/EmailStatusEventService.kt]: Use `Instant.toTimestamp()`

File: tachikoma-database/src/main/kotlin/com/sourceforgery/tachikoma/database/dao/EmailStatusEventDAOImpl.kt [hunk 10]
```diff
@@ -1,13 +1,25 @@
...
         private val ebeanServer: EbeanServer
 ) : EmailStatusEventDAO {
+
     override fun save(emailStatusEventDBO: EmailStatusEventDBO) = ebeanServer.save(emailStatusEventDBO)
+
+    override fun getEventsAfter(accountDBO: AccountDBO, instant: Instant): List<EmailStatusEventDBO> {
+        return ebeanServer
+                .find(EmailStatusEventDBO::class.java)
+                .where()
+                .eq("email.transaction.authentication.account", accountDBO)
```
Comment (robin-wallberg) [inline tachikoma-database/src/main/kotlin/com/sourceforgery/tachikoma/database/dao/EmailStatusEventDAOImpl.kt]: .id, accountId.accountId?

File: tachikoma-database/src/test/kotlin/com/sourceforgery/tachikoma/database/dao/EmailStatusEventDAOSpec.kt [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+                    emailStatus = EmailStatus.DELIVERED
+            )
+
+            val eventsTimeLimit = Instant.now().plus(2, ChronoUnit.DAYS)
+
+            val emailStatusEvents = emailStatusEventDAO.getEventsAfter(authentication1.account, eventsTimeLimit)
+
+            assertEquals(0, emailStatusEvents.size)
+        }
+    }
```
Comment (robin-wallberg) [inline tachikoma-database/src/test/kotlin/com/sourceforgery/tachikoma/database/dao/EmailStatusEventDAOSpec.kt]: Add test for old event (set dateChanged)

File: tachikoma-database/src/test/kotlin/com/sourceforgery/tachikoma/database/dao/EmailStatusEventDAOSpec.kt [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+            val authentication2 = createAuthentication("example.com")
+            createEmailStatusEvent(
+                    authentication = authentication2,
+                    from = Email("from@example.com"),
+                    recipient = Email("recipient@example.com"),
+                    emailStatus = EmailStatus.DELIVERED
+            )
+
+            val eventsTimeLimit = Instant.now().minus(2, ChronoUnit.DAYS)
```
Comment (robin-wallberg) [inline tachikoma-database/src/test/kotlin/com/sourceforgery/tachikoma/database/dao/EmailStatusEventDAOSpec.kt]: Use clock, also below

File: tachikoma-frontend-api-proto/src/main/proto/com/sourceforgery/tachikoma/grpc/admin/emailstatusevent/emailstatusevent.proto [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+service EmailStatusEventService {
+
+    // Returns all email status events the last 7 days
+    rpc getEmailStatusEvents (GetEmailStatusEventsFilter) returns (stream EmailNotification);
+}
+
+// Filter for retrieving historic e-mail status events
+message GetEmailStatusEventsFilter {
+    // Only retrieve events that are maximum this amount of days old
+    int32 daysOld = 101;
```
Comment (robin-wallberg) [inline tachikoma-frontend-api-proto/src/main/proto/com/sourceforgery/tachikoma/grpc/admin/emailstatusevent/emailstatusevent.proto]: Use timestamp

File: tachikoma-grpc/src/main/kotlin/com/sourceforgery/tachikoma/emailstatusevent/EmailStatusEventService.kt [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+        private val authentication: Authentication,
+        private val authenticationDAO: AuthenticationDAO,
+        private val emailStatusEventDAO: EmailStatusEventDAO
+) {
+    fun getEmailStatusEvents(request: GetEmailStatusEventsFilter, responseObserver: StreamObserver<EmailNotification>) {
+
+        authentication.requireFrontend()
+        val authenticationDBO = authenticationDAO.getActiveById(authentication.authenticationId)!!
+
+        val timeFromNow = Instant.now().minus(request.daysOld.toLong(), ChronoUnit.DAYS)
```
Comment (robin-wallberg) [inline tachikoma-grpc/src/main/kotlin/com/sourceforgery/tachikoma/emailstatusevent/EmailStatusEventService.kt]: Use timestamp, never Instant.now

File: tachikoma-grpc/src/main/kotlin/com/sourceforgery/tachikoma/emailstatusevent/EmailStatusEventServiceGrpcImpl.kt [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+internal class EmailStatusEventServiceGrpcImpl
+@Inject
+private constructor(
+        private val emailStatsEventService: EmailStatusEventService,
+        private val grpcExceptionMap: GrpcExceptionMap
+) : EmailStatusEventServiceGrpc.EmailStatusEventServiceImplBase() {
+
+    override fun getEmailStatusEvents(request: GetEmailStatusEventsFilter, responseObserver: StreamObserver<EmailNotification>) {
+        try {
+            emailStatsEventService.getEmailStatusEvents(request, responseObserver)
```
Comment (robin-wallberg) [inline tachikoma-grpc/src/main/kotlin/com/sourceforgery/tachikoma/emailstatusevent/EmailStatusEventServiceGrpcImpl.kt]: onComplete here

File: tachikoma-grpc/src/main/kotlin/com/sourceforgery/tachikoma/emailstatusevent/EmailStatusEventService.kt [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+
+        emailStatusEventDAO.getEventsAfter(authenticationDBO.account, timeFromNow)
+                .forEach {
+                    responseObserver.onNext(
+                            getEmailNotification(it)
+                    )
+                }
+        responseObserver.onCompleted()
```
Comment (robin-wallberg) [inline tachikoma-grpc/src/main/kotlin/com/sourceforgery/tachikoma/emailstatusevent/EmailStatusEventService.kt]: Move

**Pull request merged**